### PR TITLE
 [aes_dv] Updated AES to test DUT with back 2 back transactions

### PR DESCRIPTION
**Pull request by rasmus-madsen**
This PR is the first in a series of PRs that will update AES DV behavior and functionality as we progress towards V2.
in this PR I added random b2b variable to set blocking in csr sequence 
The back to back behavior is randomized during environment configuration and not changed during simulation.

this PR also instantiates the TL config item to make TL-UL delays configurable
this is randomized in configuration as host resposiveness
A problem exists in manual mode when responsiveness is very slow- which is fixed in a PR that will follow this.

### CODE REVIEW

File: hw/ip/aes/dv/env/aes_env_cfg.sv
Comment (tunghoang290780) [inline hw/ip/aes/dv/env/aes_env_cfg.sv:87]: nit: require spaces on both side of `-`

File: hw/ip/aes/dv/env/aes_env_cfg.sv
Comment (tunghoang290780) [inline hw/ip/aes/dv/env/aes_env_cfg.sv:113]: nits: remove empty lines 95, 113

File: hw/ip/aes/dv/env/aes_scoreboard.sv
Comment (tunghoang290780) [inline hw/ip/aes/dv/env/aes_scoreboard.sv:470]: nits: extra space after the comma

Comment (tunghoang290780) [inline hw/ip/aes/dv/env/aes_scoreboard.sv:470]:
```suggestion
          `uvm_fatal(`gfn, $sformatf(" # %d  \n\t %s \n", message_cnt, txt))
```

File: hw/ip/aes/dv/env/aes_env_cfg.sv
Comment (tunghoang290780) [inline hw/ip/aes/dv/env/aes_env_cfg.sv:113]:
```suggestion
```

File: hw/ip/aes/dv/env/aes_env_cfg.sv
Comment (tunghoang290780) [inline hw/ip/aes/dv/env/aes_env_cfg.sv:95]:
```suggestion
```

File: hw/ip/aes/dv/env/aes_env_cfg.sv
Comment (tunghoang290780) [inline hw/ip/aes/dv/env/aes_env_cfg.sv:87]:
```suggestion
                                                    1 :/ (100 - use_c_model_pct) }; }
```

File: hw/ip/aes/dv/env/aes_scoreboard.sv
Comment (tunghoang290780) [inline hw/ip/aes/dv/env/aes_scoreboard.sv:469]:
```suggestion
    n, msg.output_msg[n], msg.predicted_msg[n] )};
```

File: hw/ip/aes/dv/env/aes_scoreboard.sv
Comment (tunghoang290780) [inline hw/ip/aes/dv/env/aes_scoreboard.sv:460]:
```suggestion
    i, msg.input_msg[i], msg.output_msg[i], msg.predicted_msg[i])};
```

Comment (tunghoang290780) [inline hw/ip/aes/dv/env/aes_scoreboard.sv:460]: nits: breaking lines (i.e. 460, 469) start with 4 spaces left but it is just a preference

File: hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv
Comment (tunghoang290780) [inline hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv:53]: AFAIK, blocking is set to 1'b1 by default, so you can remove it in csr_* task unless you want to do non-block (unset to 1'b0)

File: hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv
Comment (tunghoang290780) [inline hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv:87]:
```suggestion
    `uvm_info(`gfn, $sformatf("\n\t --- back to back transactions : %b", do_b2b), UVM_MEDIUM)
```

File: hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv
Comment (rasmus-madsen) [reply hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv:53]: it is there to show It is intentionally blocking

File: hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv
Comment (tunghoang290780) [inline hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv:105]: do all the `csr_wr` need to be persistent just in case there are following reads? if yes then in the last write you can use `wait_no_outstanding_access();` after the last `csr_wr `or you can add an extra input argument do_persist (by default set to 1'b0') in the task definition then use `if (do_persist)  wait_no_outstanding_access();` for more flexibility

File: hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv
Comment (tunghoang290780) [inline hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv:166]:
```suggestion
      if (item.mode != AES_ECB) write_iv(item.iv, item.do_b2b);
```

File: hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv
Comment (tunghoang290780) [inline hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv:196]: Line 196, 271: 4 spaces left

File: hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv
Comment (tunghoang290780) [inline hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv:270]:
```suggestion
    `uvm_info(`gfn, $sformatf("\n\t ----| Configuring the DUT in the following order:  %s", txt),
```

File: hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv
Comment (tunghoang290780) [inline hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv:267]:
```suggestion
       "data_in_3": csr_wr(.csr(ral.data_in_3), .value(data[3][31:0]));
```

File: hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv
Comment (tunghoang290780) [inline hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv:197]:
```suggestion
    for (int n = 0; n < msg_item.message_length - 15; n += 16 ) begin
```

File: hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv
Comment (tunghoang290780) [inline hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv:297]:
```suggestion
```

File: hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv
Comment (tunghoang290780) [inline hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv:359]:
```suggestion
           if (status[3] && (aes_item_queue.size() > 0)) begin
```

File: hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv
Comment (tunghoang290780) [inline hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv:369]:
```suggestion
            num_blocks--;
```

File: hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv
Comment (tunghoang290780) [inline hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv:367]:
```suggestion
          if (status[2]) begin
```

**Review by tunghoang290780 — COMMENTED** [review]
Some nits (just my preference)

File: hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv
Comment (tunghoang290780) [inline hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv:119]:
```suggestion
    `uvm_info(`gfn, $sformatf("\n\t ----| DATA_IN_0: %h ", data[0]), UVM_HIGH)
```
data[i] is always 32 bit so you do not need [31:0] unless you want to have a short bit-width. Btw, you can use TL_DW-1:0 to replace 31:0 if prefer

File: hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv
Comment (tunghoang290780) [reply hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv:53]: that's fine if you want to emphasize the blocking feature

File: hw/ip/aes/dv/aes_sim_cfg.hjson
Comment (cindychip) [inline hw/ip/aes/dv/aes_sim_cfg.hjson:73]: Is this intended that `aes_b2b` test still uses `aes_stress_vseq`? Do you mean `aes_b2b_vseq`?

File: hw/ip/aes/dv/env/aes_scoreboard.sv
Comment (rasmus-madsen) [reply hw/ip/aes/dv/env/aes_scoreboard.sv:470]: fixed

File: hw/ip/aes/dv/env/aes_env_cfg.sv
Comment (rasmus-madsen) [reply hw/ip/aes/dv/env/aes_env_cfg.sv:113]: fixed

File: hw/ip/aes/dv/env/aes_env_cfg.sv
Comment (rasmus-madsen) [reply hw/ip/aes/dv/env/aes_env_cfg.sv:95]: I will keep this line.. but remove the comment above

File: hw/ip/aes/dv/env/aes_env_cfg.sv
Comment (rasmus-madsen) [reply hw/ip/aes/dv/env/aes_env_cfg.sv:87]: fixed

File: hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv
Comment (rasmus-madsen) [reply hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv:87]: ty fixed

File: hw/ip/aes/dv/env/aes_scoreboard.sv
Comment (rasmus-madsen) [reply hw/ip/aes/dv/env/aes_scoreboard.sv:460]: aligning with the open bracket seem reasonable to me

File: hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv
Comment (rasmus-madsen) [reply hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv:166]: fixed

File: hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv
Comment (rasmus-madsen) [reply hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv:105]: Good point but these are WO so this should not be needed.

File: hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv
Comment (rasmus-madsen) [reply hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv:196]: I like the align with the open bracket

File: hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv
Comment (rasmus-madsen) [reply hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv:270]: fixed

File: hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv
Comment (rasmus-madsen) [reply hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv:267]: fixed

File: hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv
Comment (rasmus-madsen) [reply hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv:197]: fixed

File: hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv
Comment (rasmus-madsen) [reply hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv:297]: fixed

File: hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv
Comment (rasmus-madsen) [reply hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv:359]: I will keep the inner brackets for readability

File: hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv
Comment (rasmus-madsen) [reply hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv:369]: I feel this is subject to preference

File: hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv
Comment (rasmus-madsen) [reply hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv:367]: fixed

File: hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv
Comment (rasmus-madsen) [reply hw/ip/aes/dv/env/seq_lib/aes_base_vseq.sv:119]: agreed it is legacy from a different array format I was using before. (I don't think it hurts thought)
I have aligned with your suggestion

File: hw/ip/aes/dv/aes_sim_cfg.hjson
Comment (rasmus-madsen) [reply hw/ip/aes/dv/aes_sim_cfg.hjson:73]: Hi @cindychip this intentional.
all or almost all of my tests will use the same sequence.
the only difference is the configuration which is done in the test.

This is another approach to what is generally used in OpenTitan.
Eventually, I will split up my env_cfg and my seq_cfg. 
We agreed a while back that this approach is accepted.

**Review by weicaiyang — COMMENTED** [review]
sorry for the late comments. I forgot to submit them last week.

File: hw/ip/aes/dv/aes_sim_cfg.hjson
Comment (tunghoang290780) [reply hw/ip/aes/dv/aes_sim_cfg.hjson:73]: @rasmus-madsen do you have a plan to enable line #36 in the `aes_sim_cfg.dv`? Are the `aes_stress_vseq `and `aes_stress_all_vseq ` (enable by line #32) different? I check the description of the aes_stress_vseq and it seems to me that is similar to the goals of the aes_stress_all_vseq. Maybe, I am missing some points but it is worth to check.

Comment (rasmus-madsen) [reply hw/ip/aes/dv/aes_sim_cfg.hjson:73]: @tunghoang290780 
conceptually they are very similar.
the stress_all_vseq stems from the autogenerated part of the testbench and is used for chaining sequences.
and with this approach there is just one testcase - the base. which is then morphed into whatever, based on the sequence referenced.

my approach is sort of the other way around. 
there is just one "top" sequence that is then configured from a testcase.
I think this is closer to what was originally intended by UVM.

there are some upsides to both methods.

Aes currently serves as an example of this other method. so we can weigh these pros and cons.

We can talk about these offline if you have an interest in the details

File: hw/ip/aes/dv/env/aes_env.sv
Comment (rasmus-madsen) [reply hw/ip/aes/dv/env/aes_env.sv:17]: ohh you are correct thank you for pointing this out

File: hw/ip/aes/dv/env/aes_env_cfg.sv
Comment (rasmus-madsen) [reply hw/ip/aes/dv/env/aes_env_cfg.sv:93]: it is unneeded. but it does not overwrite anything as the settings are on post randomize()
and comes later than the settings in the tl env.

File: hw/ip/aes/dv/env/aes_scoreboard.sv
Comment (rasmus-madsen) [reply hw/ip/aes/dv/env/aes_scoreboard.sv:468]: fixed

File: hw/ip/aes/dv/aes_sim_cfg.hjson
Comment (tunghoang290780) [reply hw/ip/aes/dv/aes_sim_cfg.hjson:73]: @rasmus-madsen ok, we can talk offline. I think this might be out of the scope of this PR

File: hw/ip/aes/dv/env/aes_scoreboard.sv
Comment (weicaiyang) [reply hw/ip/aes/dv/env/aes_scoreboard.sv]: I should suggest `0x%0h` rather than `0x%h`, which may have space between `0x` and hex number

it will be better if we use these to replace in all files.
```
find aes/dv/ -type f -exec sed -i s/0x%h/0x%0h/g {} +
find aes/dv/ -type f -exec sed -i s/%h/0x%0h/g {} +
find aes/dv/ -type f -exec sed -i s/%d/%0d/g {} +
```

**Review by weicaiyang — APPROVED** [review]
LGTM with a small comment

**Comment by rasmus-madsen** [comment]
@tunghoang290780 @cindychip 
do you have any more comments or could you please approve ?
I would like at least 2 approvers before I merge this

**Review by tunghoang290780 — APPROVED** [review]
Sorry, I forgot, just approved!

**Comment by rasmus-madsen** [comment]
thank you @tunghoang290780

**Pull request merged**
